src/renderers/WebGLRenderer.js

```diff
@@ -836,136 +823,9 @@ function WebGLRenderer( parameters ) {
-				return;
-
-			}
-
-		}
-
-		state.initAttributes();
-
-		var geometryAttributes = geometry.attributes;
-
-		var programAttributes = program.getAttributes();
-
-		var materialDefaultAttributeValues = material.defaultAttributeValues;
-
-		for ( var name in programAttributes ) {
-
-			var programAttribute = programAttributes[ name ];
-
-			if ( programAttribute >= 0 ) {
-
-				var geometryAttribute = geometryAttributes[ name ];
-
-				if ( geometryAttribute !== undefined ) {
-
-					var normalized = geometryAttribute.normalized;
-					var size = geometryAttribute.itemSize;
-
-					var attribute = attributes.get( geometryAttribute );
-
-					// TODO Attribute may not be available on context restore
-
-					if ( attribute === undefined ) continue;
-
-					var buffer = attribute.buffer;
-					var type = attribute.type;
-					var bytesPerElement = attribute.bytesPerElement;
-
-					if ( geometryAttribute.isInterleavedBufferAttribute ) {
-
-						var data = geometryAttribute.data;
-						var stride = data.stride;
-						var offset = geometryAttribute.offset;
-
-						if ( data && data.isInstancedInterleavedBuffer ) {
-
-							state.enableAttributeAndDivisor( programAttribute, data.meshPerAttribute );
-
-							if ( geometry.maxInstancedCount === undefined ) {
-
-								geometry.maxInstancedCount = data.meshPerAttribute * data.count;
-
-							}
-
-						} else {
-
-							state.enableAttribute( programAttribute );
-
-						}
-
-						_gl.bindBuffer( _gl.ARRAY_BUFFER, buffer );
-						_gl.vertexAttribPointer( programAttribute, size, type, normalized, stride * bytesPerElement, offset * bytesPerElement );
-
-					} else {
-
-						if ( geometryAttribute.isInstancedBufferAttribute ) {
-
-							state.enableAttributeAndDivisor( programAttribute, geometryAttribute.meshPerAttribute );
-
-							if ( geometry.maxInstancedCount === undefined ) {
-
-								geometry.maxInstancedCount = geometryAttribute.meshPerAttribute * geometryAttribute.count;
-
-							}
-
-						} else {
-
-							state.enableAttribute( programAttribute );
-
-						}
-
-						_gl.bindBuffer( _gl.ARRAY_BUFFER, buffer );
-						_gl.vertexAttribPointer( programAttribute, size, type, normalized, 0, 0 );
-
-					}
-
-				} else if ( materialDefaultAttributeValues !== undefined ) {
-
-					var value = materialDefaultAttributeValues[ name ];
-
-					if ( value !== undefined ) {
-
-						switch ( value.length ) {
-
-							case 2:
-								_gl.vertexAttrib2fv( programAttribute, value );
-								break;
-
-							case 3:
-								_gl.vertexAttrib3fv( programAttribute, value );
-								break;
-
-							case 4:
-								_gl.vertexAttrib4fv( programAttribute, value );
-								break;
-
-							default:
-								_gl.vertexAttrib1fv( programAttribute, value );
-
-						}
-
-					}
-
-				}
-
-			}
-
-		}
-
-		state.disableUnusedAttributes();
+		bindingStates.resetBindingState();
```

Comment:
(IIRC) I saw performance degradation by resetting VAO in the end of each `.renderBufferDirect()` call in some examples. I'll share the experiment later.

Reply:
This line was actually necessary to ensure that other render calls (e.g. by `WebGLSpriteRenderer`) don't render with an invalid binding state. It would actually be surprising if this call has a negative impact on performance since I've never saw it differently used.

Reply:
> This line was actually necessary to ensure that other render calls (e.g. by WebGLSpriteRenderer) don't render with an invalid binding state.

I know. I first tried the same approach but I saw a negative impact and started to think the way to remove that line. Removing that line makes the code a bit complicated. This's one of the reasons why I'm late to share my work.

Reply:
Remembering... One of the cases where I saw performance degradation is rendering many `Mesh`es which share the same geometry and material, like `performance` example.
In that case, ideally we don't need to reset for each `Mesh`. 

But in this implementation it calls `extension.bindVertexArrayOES( vao );` `extension.bindVertexArrayOES( null );` in each `.renderBufferDirect()`. These additional calls (compared to the current implement) seem to lead a negative impact.

IMO, I don't wanna accept this performance degradation. VAO is for improve the performance. Having a negative impact even on existing example doesn't make sense.

Reply:
So in my branch, I save the current VAO and check if we need to switch VAO by calling `.bindVertexArrayOES()`. 

But, as you know, no calling `extension.bindVertexArrayOES( null );` means VAO can be polluted somewhere. I found `WebGLAttribute` `create/updateBuffer` pollutes index(`gl.ELEMENT_ARRAY_BUFFER`).

So I'm trying a workaround

https://github.com/takahirox/three.js/blob/VAOImpleTrial/src/renderers/webgl/WebGLAttributes.js#L16
https://github.com/takahirox/three.js/blob/VAOImpleTrial/src/renderers/webgl/WebGLVertexArrayObjects.js#L93

but I know it isn't good design. I'm also thinking better solution. Perhaps we need further refactoring?

Reply:
@kenrussell To sum this up: We are not sure about the overhead of a single `extension.bindVertexArrayOES( null );`. Is it bad if we do this for each render item in order to ensure a consistent binding state?

Reply:
Not single `extension.bindVertexArrayOES( null );` overhead, but the set of `extension.bindVertexArrayOES( vao );` and `extension.bindVertexArrayOES( null );` overhead in the beginning and the end of rendering an item.

Reply:
Right, we need to perform both calls for each item.

Reply:
> I think it's important to have a better average performance over many examples.

I can't perfectly agree. Sharing Geometry and Material should be one of the good optimizations on memory and performance. With such an optimization, seeing a negative performance impact is bad. We should investigate more how to achieve **both** less changes and less(or no) performance degradation.

Reply:
> We should investigate more how to achieve both less changes and less(or no) performance degradation.

That's fine.

Reply:
In real-time 3D field, 60 (or more in VR) fps is a performance threshold. In average cases, probably the apps are already light enough and run on 60 fps without VAO. In practical most user don't see any good impact by introducing VAO. But they could see the performance degradation on heavy but well optimized apps.

IMO we should go more carefully with VAO. Targeting only average might not be good for VAO.

Reply:
I just wanted to point out that we need to find a compromise between performance a the complexity of the implementation.